Several bug fixes for request manager and major rework of IBD using the request manager

ptschip opened the pull request.

Commit:
Re-able check for end of iterator and general cleanup

Commit:
Prevent the removal of the block hash from request manager too early

When an Xthin was received the block was being indicated as recieved
before it actually was.  It was possible that a re-request had to be
made which would prevent the request manager from continuing to make
requests while the xthin could have been delayed indefinitely on it's
re-request.  Also, this fix correctly accounts for bytes received when
an Xpedited block beats and Xthin and previously the Xthin bytes were
being double counted.

Commit:
Missing cs_main lock on mapOrphanTransactions in requestmanager

Commit:
Creating new Critical Section for Orphan Cache

Removing the dependency of the orphan cache on cs_main to
streamline the request manager when requesting thinblocks.

Commit:
General Code cleanup for readability

Commit:
Do not re-request objects when rate limiting is enabled

When rate limiting is turned on there is no need for the request manager
to re-request txns and blocks which will cause further congestion on
the network and then cause even more re-requests.

Commit:
gettrafficshaping LONG_MAX should be LONG_LONG_MAX

This was causing the LONG_LONG_MAX value to be returned
and displayed when we should be displaying nothing.

Commit:
Request Manager Reject Messages: simplify code and clarify reject messages

Commit:
Remove duplicate logic for processing Xthins

Commit:
Clarify log message when xthin response is beaten

A requested Xthin can be beaten not just by an expedited block
by by a re-requested xthin.  Also wanted to be clear about which
peer was doing the requesting, it appeared that the peer doing the
requesting was actually the one that had beaten the first requester.

ptschip commented:
taking this out of WIP

Commit:
Prevent thinblock from being processed when we already have it

Do not continue processing an xblocktx response if we already
have received the thinblock.  This can happen when expedited
requesting has been turned on.

Commit:
A few more locks added for cs_orphancache

Commit:
Fix IBD hang issue by using request manager

IBD will often hang and is a longstanding issue. By using
the request manager we can re-request blocks that do not return
do to hung nodes, slow nodes or disconnects.

Commit:
Auto adjust IBD download settings depending on reponse time

Move the MAX_BLOCKS_IN_TRANSIT_PER_PEER and BLOCK_DOWNLOAD_WINDOW
up and down depending on average block response time for the last
20 blocks received during IBD.

Commit:
Remove block pacer in request manager when requesting blocks

This helps to improve performance dramatically when doing IBD on
the first part of the chain.

Commit:
Filter out the nodes with above average ping times during IBD

If the ping time for a peer is greater than 2 times the standard
deviation plus the average ping time for all valid full nodes,
then do not make block download requests
from that peer during IBD.  This dramatically boost IBD performance
when one or more very slow nodes are connected.

Commit:
take out one sendexpedited() as not necessary

Commit:
Add more detailed logging when expedited connect not possible

ptschip commented:
@gandrewstone  this one is also ready for merging.

gandrewstone commented on src/main.cpp:
```diff
@@ -3726,7 +3773,6 @@ bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, c
       }
 
     LogPrint("thin", "Processing new block %s from peer %s (%d).\n", pblock->GetHash().ToString(), pfrom ? pfrom->addrName.c_str():"myself",pfrom ? pfrom->id: 0);
-    if (!IsInitialBlockDownload()) SendExpeditedBlock(*pblock,pfrom);
```
why is the expedited block send deleted here?

gandrewstone replied:
Unfortunately it needs to be here to catch blocks coming from miners and from the relay network -- basically from the JSON interface, IIRC.  I don't like the double check either though.  I'm going to put this back in and then we can talk about what to do in a different PR.  That way we can get everything else in this PR merged.

gandrewstone replied:
yes that is better...

gandrewstone commented on src/main.h:
```diff
@@ -77,17 +77,20 @@ static const int MAX_SCRIPTCHECK_THREADS = 16;
 /** Number of blocks that can be requested at any given time from a single peer. */
-static const int MAX_BLOCKS_IN_TRANSIT_PER_PEER = 16;
+//static const int MAX_BLOCKS_IN_TRANSIT_PER_PEER = 16;
+static int MAX_BLOCKS_IN_TRANSIT_PER_PEER = 1; // BU request manager just needs one per peer
```
Comment is misleading, you are adjusting this dynamically based on block DL perf, right?  If so I will fix during merge.

ptschip replied:
The comment should be " BU request manager just needs one per peer to
start with"    we want to start with a low number in case we request
from a slow peer right at the beginning.  as the blocks start coming in
the value gets adjusted upward as needed.

09/2016 3:28 AM, gandrewstone wrote:

> In src/main.h
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79#discussion_r77450097:
> 
> > @@ -77,17 +77,20 @@ static const int MAX_SCRIPTCHECK_THREADS = 16;
> >  /*\* -par default (number of script-checking threads, 0 = auto) _/
> >  static const int DEFAULT_SCRIPTCHECK_THREADS = 0;
> >  /_\* Number of blocks that can be requested at any given time from a single peer. */
> > -static const int MAX_BLOCKS_IN_TRANSIT_PER_PEER = 16;
> > +//static const int MAX_BLOCKS_IN_TRANSIT_PER_PEER = 16;
> > +static int MAX_BLOCKS_IN_TRANSIT_PER_PEER = 1; // BU request manager just needs one per peer
> 
> Comment is misleading, you are adjusting this dynamically based on
> block DL perf, right? If so I will fix during merge.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79/files/efaaa2474cc017ebc1c84fa3748b50c293bff5c7#r77450097,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AMRF0L-GQJD1uxSNLmOl5gdPABhDatGLks5qmp1MgaJpZM4JokbP.

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -102,15 +110,11 @@ void CRequestManager::AskFor(const CInv& obj, CNode* from, int priority)
-      else  // existing
-	{
-	}
-
```
This is my shorthand for "I have considered this case and there's nothing to do"...

gandrewstone replied:
I'm good with that -- I'll add the comment

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -315,13 +315,13 @@ void RequestBlock(CNode* pfrom, CInv obj)
+  if (IsChainNearlySyncd()) // only download headers if we're not doing IBD.  The IBD process will take care of it's own headers.
   {
-    LogPrint("net", "getheaders (%d) %s to peer=%d\n", pindexBestHeader->nHeight, obj.hash.ToString(), pfrom->id);  
+    LogPrint("thin", "getheaders (%d) %s to peer=%d\n", pindexBestHeader->nHeight, obj.hash.ToString(), pfrom->id);  
```
this GETHEADERS req is not thin block specific...

ptschip replied:
true, i was trying to see the messages during debugging without having
to put full net debug on...i can change it back.

On 04/09/2016 3:36 AM, gandrewstone wrote:

> In src/requestManager.cpp
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79#discussion_r77450214:
> 
> >    {
> > -    LogPrint("net", "getheaders (%d) %s to peer=%d\n", pindexBestHeader->nHeight, obj.hash.ToString(), pfrom->id);  
> > -    LogPrint("thin", "getheaders (%d) %s to peer=%d\n", pindexBestHeader->nHeight, obj.hash.ToString(), pfrom->id);  
> 
> this GETHEADERS req is not thin block specific...
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79/files/efaaa2474cc017ebc1c84fa3748b50c293bff5c7#r77450214,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AMRF0MvQJYV7-a8r0Ds-Qnyy-BoJ0vcoks5qmp8mgaJpZM4JokbP.

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -394,67 +400,77 @@ void CRequestManager::SendRequests()
-		  if (item.lastRequestTime && IsChainNearlySyncd())  // if this is positive, we've requested at least once
+                  // If item.lastRequestTime is true then we've requested at least once and we'll try a re-request if the following conditions are met:
+                  //     The chain must be almost syncd and traffic shaping must not be turned on
+		  if (item.lastRequestTime && !IsTrafficShapingEnabled())
```
why no rerequest if traffic shaping?  sure shaping could slow down the receipt of blocks, but clearly blocks could still be dropped if you are shaping or not.  It seems like we have an interaction with traffic shaping that could be problematic -- that is, there exists pathological configurations where your own traffic shaping causes block rerequests.  However turning off rerequesting if traffic shaping is turned on will cause runtime issues for drops during normal configuration.  

I think that the predicate we are looking for here is more like "ReceiveBufferRecentlyEmpty()"

ptschip replied:
@gandrewstone  I'm rethinking this one...i think maybe we could do something like this:  re-enable re-requests when traffic shaping is on but bump the re-request timeout to a higher value like 30 seconds...and in the future calculate a useful timeout based on network throughput or traffic shaping settings etc...what do you think?

gandrewstone commented on src/requestManager.cpp:
```diff
@@ -522,39 +539,61 @@ void CRequestManager::SendRequests()
-                        }
-                      cs_objDownloader.lock();
-		      }
-                    if (0)  // SAME NODE RETRY DISABLED: Won't help in a TCP connection anyway: 
```
WRT removal of "if (0)" and subsequent logic: I want to preserve my thought process here and keep the code logically correct in case we move to multiple transports...

ptschip replied:
I'm not sure I agree with leaving our thought processes in the main code
branch, it makes the code more difficult to read through a complex
section of code IMO.  Other than //TODO comments, I believe those are
things we should have in our own private branches that we're working
on.  But it's up to you, i'm not going to make a big thing out of it...

.On 04/09/2016 3:48 AM, gandrewstone wrote:

> In src/requestManager.c
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79#discussion_r77450382
> 
> > @@ -522,39 +539,61 @@ void CRequestManager::SendRequests()
> >             {
> >                      if (1)
> >                        {
> > -                      cs_objDownloader.unlock();
> > -                      LOCK(next.node->cs_vSend);
> > -             // from->AskFor(item.obj); basically just shoves the req into mapAskFor
> > -                      if (!item.lastRequestTime || (item.lastRequestTime && IsChainNearlySyncd()))
> > -                        {
> > -                          next.node->mapAskFor.insert(std::make_pair(now, item.obj));
> > -                          item.outstandingReqs++;
> > -                 item.lastRequestTime = now;
> > -                        }
> > -                      cs_objDownloader.lock();
> > -             }
> > -                    if (0)  // SAME NODE RETRY DISABLED: Won't help in a TCP connection anyway: 
> 
> WRT removal of "if (0)" and subsequent logic: I want to preserve my
> thought process here and keep the code logically correct in case we
> move to multiple transports...
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79/files/efaaa2474cc017ebc1c84fa3748b50c293bff5c7#r77450382,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AMRF0AKDSr0sDaLrSA5Q8n1in31fPA3oks5qmqHqgaJpZM4JokbP.

ptschip commented on src/main.cpp:
```diff
@@ -3726,7 +3773,6 @@ bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, c
       }
 
     LogPrint("thin", "Processing new block %s from peer %s (%d).\n", pblock->GetHash().ToString(), pfrom ? pfrom->addrName.c_str():"myself",pfrom ? pfrom->id: 0);
-    if (!IsInitialBlockDownload()) SendExpeditedBlock(*pblock,pfrom);
```
this already gets called when we receive a normal block or xthin, why do
it twice?

On 04/09/2016 3:17 AM, gandrewstone wrote:

> In src/main.cpp
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79#discussion_r77449973:
> 
> > @@ -3726,7 +3773,6 @@ bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, c
> >        }
> > 
> > ```
> >  LogPrint("thin", "Processing new block %s from peer %s (%d).\n", pblock->GetHash().ToString(), pfrom ? pfrom->addrName.c_str():"myself",pfrom ? pfrom->id: 0);
> > ```
> > -    if (!IsInitialBlockDownload()) SendExpeditedBlock(*pblock,pfrom);
> 
> why is the expedited block send deleted here?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79/files/efaaa2474cc017ebc1c84fa3748b50c293bff5c7#r77449973,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AMRF0BF51XDNqDG5uOxDbzm8hmdYe8lNks5qmpq8gaJpZM4JokbP.

ptschip commented on src/requestManager.cpp:
```diff
@@ -102,15 +110,11 @@ void CRequestManager::AskFor(const CInv& obj, CNode* from, int priority)
-      else  // existing
-	{
-	}
-
```
IMO if that's the case we should leave them in but comment them out with
a short explanation such as // else - nothing to do 

On 04/09/2016 3:32 AM, gandrewstone wrote:

> In src/requestManager.cpp
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79#discussion_r77450168:
> 
> > @@ -102,15 +110,11 @@ void CRequestManager::AskFor(const CInv& obj, CNode\* from, int priority)
> >       pendingTxns+=1;
> >       // all other fields are zeroed on creation
> >     }
> > -      else  // existing
> > -   {
> > 
> > ## \-   }
> 
> This is my shorthand for "I have considered this case and there's
> nothing to do"...
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79/files/efaaa2474cc017ebc1c84fa3748b50c293bff5c7#r77450168,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AMRF0Lf9AtPRbO6UUzFAKKJQLWPtSE30ks5qmp5agaJpZM4JokbP.

ptschip commented on src/requestManager.cpp:
```diff
@@ -394,67 +400,77 @@ void CRequestManager::SendRequests()
-		  if (item.lastRequestTime && IsChainNearlySyncd())  // if this is positive, we've requested at least once
+                  // If item.lastRequestTime is true then we've requested at least once and we'll try a re-request if the following conditions are met:
+                  //     The chain must be almost syncd and traffic shaping must not be turned on
+		  if (item.lastRequestTime && !IsTrafficShapingEnabled())
```
My reasoning here was that if traffic shaping is on then and we lose a
block we'll still get it eventually when we hit the block download
timeout...as for txns it's no big deal IMO if we lose a few with traffic
shaping on.  It's obviously not a perfect solution but it does prevent a
lot of unecessary downloading of txns and blocks when the shaping is on,
which is exactly what the person who has shaping on is trying to
prevent.   That said, when doing IBD, the re-reqeusts will still happen
for blocks  to make sure the IBD process doesn't get hung up.

how can receivebufferrecentlyempty() help?

On 04/09/2016 3:45 AM, gandrewstone wrote:

> In src/requestManager.cpp
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79#discussion_r77450336:
> 
> > ```
> >           }
> >       }
> >         }
> > 
> >       if (next.node != NULL )
> >     {
> > ```
> > -         if (item.lastRequestTime && IsChainNearlySyncd())  // if this is positive, we've requested at least once
> > -                  // If item.lastRequestTime is true then we've requested at least once and we'll try a re-request if the following conditions are met:
> > -                  //     The chain must be almost syncd and traffic shaping must not be turned on
> > -         if (item.lastRequestTime && !IsTrafficShapingEnabled())
> 
> why no rerequest if traffic shaping? sure shaping could slow down the
> receipt of blocks, but clearly blocks could still be dropped if you
> are shaping or not. It seems like we have an interaction with traffic
> shaping that could be problematic -- that is, there exists
> pathological configurations where your own traffic shaping causes
> block rerequests. However turning off rerequesting if traffic shaping
> is turned on will cause runtime issues for drops during normal
> configuration.
> 
> I think that the predicate we are looking for here is more like
> "ReceiveBufferRecentlyEmpty()"
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79/files/efaaa2474cc017ebc1c84fa3748b50c293bff5c7#r77450336,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AMRF0MVnGdxYraRb5yway6IMOm0MCBx7ks5qmqE4gaJpZM4JokbP.

Commit:
Fix potential deadlock with cs_orphancache

Commit:
Clarify explantion for comments

Commit:
Revert LogPrint statement back to "net" instead of "thin"

Commit:
Renable re-requests when traffic shaping is enabled

Instead of disabling re-request we change the re-try interval
from 5 seconds to 30 seconds for both blocks and transactions.

Also transactions will can not be re-requested unti the
chainisnearlysyncd() returns true however blocks can always be
re-requested at any time because we need that functionaly during IBD.

ptschip commented on src/main.cpp:
```diff
@@ -3726,7 +3773,6 @@ bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, c
       }
 
     LogPrint("thin", "Processing new block %s from peer %s (%d).\n", pblock->GetHash().ToString(), pfrom ? pfrom->addrName.c_str():"myself",pfrom ? pfrom->id: 0);
-    if (!IsInitialBlockDownload()) SendExpeditedBlock(*pblock,pfrom);
```
oh i see, but maybe you could change it to   if (IsChainNearlySyncd())
instead?

On 12/09/2016 2:46 PM, gandrewstone wrote:

> In src/main.cpp
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79#discussion_r78461029:
> 
> > @@ -3726,7 +3773,6 @@ bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, c
> >        }
> > 
> > ```
> >  LogPrint("thin", "Processing new block %s from peer %s (%d).\n", pblock->GetHash().ToString(), pfrom ? pfrom->addrName.c_str():"myself",pfrom ? pfrom->id: 0);
> > ```
> > -    if (!IsInitialBlockDownload()) SendExpeditedBlock(*pblock,pfrom);
> 
> Unfortunately it needs to be here to catch blocks coming from miners
> and from the relay network -- basically from the JSON interface, IIRC.
> I don't like the double check either though. I'm going to put this
> back in and then we can talk about what to do in a different PR. That
> way we can get everything else in this PR merged.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79/files/efaaa2474cc017ebc1c84fa3748b50c293bff5c7#r78461029,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AMRF0Nm5yCxl7NTybDOQ3J96oIw5aUJ8ks5qpcgkgaJpZM4JokbP.
